core/engine/src/builtins/map/mod.rs

```diff
@@ -549,6 +555,111 @@ impl Map {
         MapIterator::create_map_iterator(this, PropertyNameKind::Value, context)
     }
 
+    /// `Map.prototype.getOrInsert(key, value)`
+    ///
+    /// Given a key and a value, the getOrInsert method will return the existing value if it exists.
+    /// Otherwise insert the provided default value and return that value.
+    ///
+    /// More information:
+    ///  - [Upsert proposal reference][spec]
+    ///  - [MDN documentation][mdn]
+    ///
+    /// [spec]: https://tc39.es/proposal-upsert/#sec-map.prototype.getOrInsert
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/getOrInsert
+    pub(crate) fn get_or_insert(
+        this: &JsValue,
+        args: &[JsValue],
+        _: &mut Context,
+    ) -> JsResult<JsValue> {
+        let key = args.get_or_undefined(0);
+        let value = args.get_or_undefined(1);
+
+        // 1. Let M be the this value.
+        let map_obj = this.as_object();
+        // 2. Perform ? RequireInternalSlot(M, [[MapData]]).
+        let mut map = map_obj
+            .as_ref()
+            .and_then(|obj| obj.downcast_mut::<OrderedMap<JsValue>>())
+            .ok_or_else(|| JsNativeError::typ().with_message("`this` is not a Map"))?;
+
+        // 3. Set key to CanonicalizeKeyedCollectionKey(key).
+        let key = match key.variant() {
+            // If key is -0𝔽, set key to +0𝔽.
+            JsVariant::Float64(r) if r.is_zero() => JsValue::new(0),
+            _ => key.clone(),
+        };
```

Comment:
We have this implemented: [`canonicalize_keyed_collection_value`](https://github.com/boa-dev/boa/blob/3bf25a692bc793a3dfaf501c935891eeb9fb8d24/core/engine/src/builtins/set/mod.rs#L127). Maybe we could move that function to a utility since it'd be used outside of set.